src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -85,26 +84,67 @@ public Task RunAsync(ActionRunStage stage)
                 actionRunner.Stage = stage;
                 actionRunner.Condition = aStep.Condition;
 
-                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, location, Environment);
-
+                var step = ExecutionContext.RegisterNestedStep(actionRunner, inputsData, Environment);
                 InitializeScope(step);
 
-                location++;
+                // Add all steps to the Composite StepsRunner
+                // Follows similar logic to how JobRunner invokes the StepsRunner for job steps!
+                compositeSteps.Add(step);
             }
 
-            // Create a step that handles all the composite action steps' outputs
-            Pipelines.ActionStep cleanOutputsStep = new Pipelines.ActionStep();
-            cleanOutputsStep.ContextName = ExecutionContext.ContextName;
-            // Use the same reference type as our composite steps.
-            cleanOutputsStep.Reference = Action;
+            try
+            {
+                // This is where we run each step.
+                await RunStepsAsync(compositeSteps, ExecutionContext);
+
+                // This is where we set the outputs.
+                // All steps have the same scope for their ExecutionContext/
+                // So we can use one of them so that we can get the right "outputs" attribute.
+                if (compositeSteps.Count > 0)
+                {
+                    // TODO: Find better solution. 
+                    HandleOutput(compositeSteps[0].ExecutionContext);
```

Comment:
Just had an idea...

Let's use the `ExecutionContext`. Technically speaking, calculating the outputs is part of the outer step (this handler) and therefore (arguably) makes sense to use to this handler's execution context.

Before evaluating the outputs, you will need to set `ExecutionContext.ExpressionValues["inputs"] = inputsData;` and `ExecutionContext.ExpressionValues["steps"] = StepsContext.GetScope(fullyQualifiedContextName);`

Reply:
Genius

Reply:
+1

Reply:
Changed to:
```
                // Get the pointer of the correct "steps" object and pass it to the ExecutionContext so that we can process the outputs correctly
                // This will always be the same for every step so we can pull this from the first step if it exists
                var stepExecutionContext = compositeSteps.Count > 0 ? compositeSteps[0].ExecutionContext : null;
                ExecutionContext.ExpressionValues["inputs"] = inputsData;
                ExecutionContext.ExpressionValues["steps"] = stepExecutionContext.StepsContext.GetScope(stepExecutionContext.ScopeName);

                HandleOutput();
```